### PR TITLE
ARUHA-164: validate partition keys

### PR DESCRIPTION
**Pull request by v1ctor**

### CODE REVIEW

File: src/main/java/org/zalando/nakadi/service/EventTypeService.java [hunk 1]
```diff
@@ -180,52 +181,50 @@ private void validateSchemaChange(final EventType eventType, final EventType exi
         }
     }
 
-    private void validatePartitionKeys(final EventType eventType) throws InvalidEventTypeException {
```
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/service/EventTypeService.java]: The problem with removing this toggle is that currently it's impossible to change schemas but it's possible to change partition keys. But after merging this code it will be impossible to change partition keys if they are not defined in the schema.

But anyway, I think if we notify users properly, there should be no major consequences.

Comment (v1ctor) [inline src/main/java/org/zalando/nakadi/service/EventTypeService.java]: @rcillo This method has been moved below validaeSchema, and there is toggle :)

File: src/main/java/org/zalando/nakadi/service/EventTypeService.java [hunk 2]
```diff
@@ -180,52 +181,50 @@ private void validateSchemaChange(final EventType eventType, final EventType exi
...
     }
 
     private void assignTopic(final EventType eventType) {
         eventType.setTopic(uuidGenerator.randomUUID().toString());
     }
 
-    private boolean hasReservedField(final JSONObject schemaAsJson, final String field) {
-        return schemaAsJson.optJSONObject("properties") != null
-                && schemaAsJson.getJSONObject("properties").has(field);
+    private String convertToJSONPointer(final String value) {
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/service/EventTypeService.java]: I think it is better to put in utility class

File: src/test/java/org/zalando/nakadi/validation/JSONSchemaValidationTest.java [hunk 3]
```diff
@@ -57,7 +57,7 @@ public void schemaValidationShouldRespectEventTypeDefinition() {
                 equalTo("#: required key [bar] not found"));
         assertThat(validator.validate(invalidEventMissingNameField).get().getMessage(), equalTo("name is required"));
         assertThat(validator.validate(nestedSchemaViolation).get().getMessage(),
-                equalTo("#: 2 schema violations found\n#/bar: 2 schema violations found\n#/bar: required key [foo]" +
+                equalTo("#: 3 schema violations found\n#/bar: 2 schema violations found\n#/bar: required key [foo]" +
```
Comment (rcillo) [inline src/test/java/org/zalando/nakadi/validation/JSONSchemaValidationTest.java]: Is this related to some other fix in the library?

File: src/main/java/org/zalando/nakadi/service/EventTypeService.java [hunk 4]
```diff
@@ -180,52 +181,50 @@ private void validateSchemaChange(final EventType eventType, final EventType exi
...
+        return value.replaceAll("\\.", "/");
```
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/service/EventTypeService.java]: I would not like to go into the details of JSON Pointer, since our previous path representation is simplified, but maybe we should have another ticket to avoid definition of properties with weird string, like https://tools.ietf.org/html/rfc6901#section-5

Because if we allow that, then this transformation could fail.

**Comment by rcillo** [comment]
👍

**Comment by rcillo** [comment]
Please, add a test case with "$ref"

File: src/test/java/org/zalando/nakadi/validation/JSONSchemaValidationTest.java (hunk 3 above)
Comment (v1ctor) [inline src/test/java/org/zalando/nakadi/validation/JSONSchemaValidationTest.java]: Yes.

**Comment by v1ctor** [comment]
Could you please describe in more details?

**Comment by rcillo** [comment]
@v1ctor  partition keys should work with schemas that have "definitions" and thus "$ref" fields.

**Comment by v1ctor** [comment]
Ok, I can. But I have written this tests in library would you like me to write the same tests in nakadi?

**Comment by rcillo** [comment]
If that's the case, then I see no need for that.

**Comment by rcillo** [comment]
:shipit:
